API: Use unsigned byte-wise comparison to fix UUID comparison bug

bodduv opened the pull request:
Fixes #14216

The problem is described in the issue as well as in the [dev mailing list post]( https://lists.apache.org/thread/6r4fkho84z4240fpk4kl5bnlf542n0pm).

The changes proposed in the PR take a simple approach to fix the comparisons bug. This directly addresses correctness issues at both `ManifestEntry` as well as `ManifestFile` filtering. Note that the changes make UUID values comparison forward-compatible, but compliant with UUID RFCs.

Additionally, the changes here maintain backward compatibility with files written before RFC 4122/9562 compliant comparison was implemented. This is accomplished by evaluating predicate expression once with RFC compliant comparator and again with signed comparator. If either of the evaluations is true, then the ManifestEntry or ManifestFile is filtered-in. This conservative approach ensures files are not skipped incorrectly.

bodduv committed:
API: Use unsigned byte-wise comparison to fix UUID comparison bug

Java's UUID.compareTo() uses signed comparison of most significant bits
and least significant bits, which is not compliant with RFC 4122, RFC 9562.
This causes incorrect ManifestEntry and ManifestFile filtering/pruning
in the presence of UUID filters.

bodduv commented:
Tagging @nastra and @Fokko based on your previous work on UUIDs and @dimas-b as I touch `TestComparators` file you added.

dimas-b reviewed (approved):
The proposed change looks reasonable to me given the explanation in #14216... however, I cannot assess the impact on existing Iceberg clients and tables. I hope other reviewers can help with that.

nastra commented on api/src/test/java/org/apache/iceberg/expressions/TestInclusiveManifestEvaluator.java:
```diff
@@ -753,4 +769,167 @@ public void testIntegerNotIn() {
+        ManifestEvaluator.forRowFilter(equal("uuid", UUID_MIN_VALUE), SPEC, true).eval(FILE);
+    assertThat(shouldRead).as("Should read: uuid equal to lower bound").isTrue();
+
+    UUID middle = UUID.fromString("7fffffff-ffff-ffff-7fff-ffffffffffff");
```
```suggestion
    UUID between = UUID.fromString("7fffffff-ffff-ffff-7fff-ffffffffffff");
```

nastra commented on api/src/test/java/org/apache/iceberg/expressions/TestInclusiveManifestEvaluator.java:
```diff
@@ -753,4 +769,167 @@ public void testIntegerNotIn() {
     assertThat(shouldRead).as("Should read: notIn on no nulls column").isTrue();
   }
+
+  /** Tests UUID equality filter using byte-order comparison against partition bounds. */
```
nit: I believe all of these comments can be removed, since it's obvious from the method naming

nastra commented on api/src/test/java/org/apache/iceberg/expressions/TestInclusiveMetricsEvaluator.java:
```diff
@@ -119,24 +132,29 @@ public class TestInclusiveMetricsEvaluator {
-              11, toByteBuffer(Types.FloatType.get(), Float.NaN),
-              12, toByteBuffer(Types.DoubleType.get(), Double.NaN),
-              14, toByteBuffer(Types.StringType.get(), "")),
+          ImmutableMap.<Integer, java.nio.ByteBuffer>builder()
```
why is the change from `ImmutableMap.of(..)` to the builder needed here?

nastra commented:
overall the change LGTM, just left a few minor comments

bodduv commented on api/src/test/java/org/apache/iceberg/expressions/TestInclusiveMetricsEvaluator.java:
```diff
@@ -970,4 +988,191 @@ public void testNotNullInNestedStruct() {
+  }
+
+  /**
+   * Tests UUID greater-than-or-equal filter using byte-order comparison against data file metrics.
```
removed it

bodduv commented:
Thank you for the comment @pvary 
> * We have a table with an UUID column
> * We inserted 2 rows to the table with UUID_MIN and UUID_MAX with Java Iceberg 1.10.0, and calculated column stats (min=UUID_MAX, max=UUID_MIN)

It matter how a query engine prepares min, max values for UUID columns to hand them over for writing manifest file and manifest lists. Some engines could use min and max values as prepared by Parquet Java (which is RFC compliant) during writes.

> * We run a query which filter on UUID_MIDDLE.
>   
>   * I expect that the metadata filtering will return the new file (UUID_MAX < UUID_MIDDLE < UUID_MIN), and we will find the row
> 
> Am I correct, that after the upgrade the metadata filtering will skip the new file (UUID_MIDDLE < UUID_MAX) - filtered out by the wrong min value?

Yes, if the min and max metrics persisted in manifest file and manifest list are constructed using the faulty non-RFC compliant UUID comparisons, then yes we would not be able to read the new file back with such a filter (on UUID column) after upgrading. What is even more problematic (evident in my testing) that even an equality filter `uuid_col = ...` will leave out records that are supposed to be returned. Note that with a full table scan we will be read the new file.

A remedy would be to migrate the table (doing a full table scan) to rewrite metrics accurately.

Note: This issue is only in Java implementation of the spec. Go, Rust, Cpp implementations are RFC compliant making the bug more severe. I.e., If the same table is read with a filter using Go implementation, it produced correct, but different records than what Java implementation produces.

amogh-jahagirdar reviewed (commented):
I echo @pvary concern that this is a pretty significant behavior change, as the ordering won't be stable for UUIDs between versions. I'm digging into the RFCs exact language but I also came across https://stackoverflow.com/questions/79489893/does-the-uuid-comparison-in-java-violate-the-uuid-standard it sounds like this sorting behavior is particularly defined for UUIDV6/7 and sorting is not prescribed in V4? At the same time, the openJDK folks acknowledge that this is a bug, so I'm not sure which yet (like I said, still digging into it) https://bugs.openjdk.org/browse/JDK-7025832

In general, I think we should close this ambiguity on the sorting of UUIDs in the spec definition; as you pointed out,  implementations are inconsistent in how this is performed. Whether or not this definition should be based off the RFC or if there's a good argument to retroactively work from the Java reference implementation behavior is something I'm still thinking through, and I think we should discuss in the mailing list.

bodduv committed:
Let ManifestFile and ManifestEntry to be read if filter expression passes either RFC or non-RFC compliant UUID comparator

pvary commented on api/src/main/java/org/apache/iceberg/expressions/ExpressionUtil.java:
```diff
@@ -744,4 +745,200 @@ private static PartitionSpec identitySpec(Schema schema, int... ids) {
+          }
+        }
+      }
+      return false;
```
nit: newline

pvary commented on api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java:
```diff
@@ -104,8 +135,22 @@ private boolean eval(ContentFile<?> file) {
+      if (useSignedUuidComparator && term.ref().type().typeId() == Type.TypeID.UUID) {
+        return (Comparator<T>) Comparators.signedUUIDs();
+      }
+      return ((BoundTerm<T>) term).comparator();
```
nit: newline

bodduv committed:
Do only a single tree traversal; address comments

bodduv commented on api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java:
```diff
@@ -86,8 +113,12 @@ private class MetricsEvalVisitor extends ExpressionVisitors.BoundVisitor<Boolean
+    private boolean useSignedUuidComparator = false;
 
-    private boolean eval(ContentFile<?> file) {
+    private boolean eval(ContentFile<?> file, Expression expression, boolean signedUuidMode) {
```
Ah I missed that, removed expression as parameter.

bodduv commented on api/src/main/java/org/apache/iceberg/expressions/ExpressionUtil.java:
```diff
@@ -744,4 +746,125 @@ private static PartitionSpec identitySpec(Schema schema, int... ids) {
+   * @return Optional containing the transformed expression if UUID bounds predicates exist, or
+   *     empty if no UUID bounds predicates were found
+   */
+  public static Optional<Expression> toSignedUUIDLiteral(Expression expr) {
```
This is related to single tree traversal change (as opposed to doing twice). `Optional.empty` denotes that no UUID bound predicate was found. Looking elsewhere looks like its not common, I moved to null checks.

bodduv committed:
Address comments; drop expression as a parameter of eval(); move to null checks

bodduv committed:
Consolidate comparatorForIn into a static method

pvary commented on api/src/main/java/org/apache/iceberg/types/Comparators.java:
```diff
@@ -251,6 +251,24 @@ public static Comparator<UUID> signedUUIDs() {
+    if (useSignedUuid && type.typeId() == Type.TypeID.UUID) {
+      return (Comparator<T>) signedUUIDs();
+    }
+    return defaultComparator;
```
nit: newline before the `return`

bodduv committed:
Fix iff; add new lines after block; fix TestInclusiveManifestEvaluator conflicts

pvary reviewed (approved):
+1 from my side.
I would love to see more eyes on it, but I think the PR is solid, and we can merge sometime next week if there are no further questions.

nastra commented on api/src/test/java/org/apache/iceberg/expressions/TestExpressionUtil.java:
```diff
@@ -1350,4 +1351,105 @@ private void assertEquals(UnboundTransform<?, ?> expected, UnboundTransform<?, ?
+    UnboundPredicate<?> pred = (UnboundPredicate<?>) result;
+    assertThat(pred.literal().value()).isEqualTo(testUuid);
+    // The literal should now use the signed comparator
+    assertThat(pred.literal()).isInstanceOf(Literals.UUIDLiteral.class);
```
we should make sure here that the signed comparator is actually being used via
```
assertThat(pred.literal())
        .isInstanceOf(Literals.UUIDLiteral.class)
        .asInstanceOf(InstanceOfAssertFactories.type(Literals.UUIDLiteral.class))
        .extracting("useSignedComparator")
        .isEqualTo(true);
```

nastra replied:
please make sure to add such a check to the other test cases in this class as well

nastra commented on api/src/test/java/org/apache/iceberg/expressions/TestExpressionUtil.java:
```diff
@@ -1350,4 +1351,105 @@ private void assertEquals(UnboundTransform<?, ?> expected, UnboundTransform<?, ?
+
+    assertThat(ExpressionUtil.toSignedUUIDLiteral(original))
+        .as("Should detect UUID lt predicate")
+        .isNotNull();
```
I feel like these can be just combined with the other tests where we end up checking that the literal actually uses the correct comparator. We should still make sure that we cover all eq/gt/lt/... cases though

bodduv committed:
Address comments for test files

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/expressions/Literals.java:
```diff
@@ -608,9 +608,24 @@ public String toString() {
+    private transient volatile Comparator<UUID> cmp;
+
     UUIDLiteral(UUID value) {
+      this(value, false);
```
Won't this change the behavior for existing library usage outside of Iceberg? Feels like this would be a breaking change? Or is this mostly unimportant from an end user perspective?

amogh-jahagirdar replied:
I think I ultimately look at this as while it's a breaking change in behavior it's probably unimportant, particularly for downstream consumers comparing literal. It already was kind of  a gap for the regular stats case for a long time, a probably bigger use case which is being fixed in this PR and handled in a compatible manner, and then for the literal case it's basically a potential issue for any consumer who is comparing on their own but that seems even more rare. So all that to say I agree with @bodduv rationale here; maybe there's some conf/property based approach but seems overkill and I'd prefer to just "rip the bandaid" off here for this case.

stevenzwu commented on api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java:
```diff
@@ -56,14 +56,25 @@ public class InclusiveMetricsEvaluator {
+  // Expression using signed UUID comparator for backward compatibility with files written
+  // prior to the introduction of RFC-compliant UUID comparisons.
+  // Null if there are no UUID predicates comparing against bounds in the expression.
+  private final Expression signedUuidExpr;
```
Instead of generating a `signedUuidExpr`, can we just change the `compartor.compare(...)` for `Literal` and `BoundTerm` to a wrapper method, which check the literal or field type and performs the additional signed comparator? I am wondering if this can reduce the size of the changes.

bodduv commented on api/src/main/java/org/apache/iceberg/expressions/Literals.java:
```diff
@@ -608,9 +608,24 @@ public String toString() {
+    // Flag to indicate which comparator to use (serializable)
+    private final boolean useSignedComparator;
+    // Transient cached comparator (reconstructed on deserialization)
+    private transient volatile Comparator<UUID> cmp;
```
Yeah I think this is not particularly important. I will drop `cmp` and keep it simple.

bodduv committed:
Drop transient cached cmp

amogh-jahagirdar reviewed (commented):
Thanks @bodduv from my side I fundamentally agree with the fix, I need a bit more time to go through all the tests.

bodduv committed:
Refactor ExpressionUtil.predicate

bodduv commented:
Thank you @amogh-jahagirdar for taking the time to review. I addressed the comments made so far.

stevenzwu commented on api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java:
```diff
@@ -74,7 +85,17 @@ public InclusiveMetricsEvaluator(Schema schema, Expression unbound, boolean case
+    }
+
+    // Always try with signed UUID comparator as a fallback. There is no reliable way to detect
+    // which comparator was used when the file's column metrics were written.
```
nit: maybe the comment can be a little more clear
```
whether signed or unsigned comparator was used when the UUID column stats were written.
```

bodduv committed:
Make MetricsEvalVisitor static; address doc related comments

stevenzwu commented on api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java:
```diff
@@ -86,8 +107,12 @@ private class MetricsEvalVisitor extends ExpressionVisitors.BoundVisitor<Boolean
+    private boolean useSignedUuidComparator = false;
 
-    private boolean eval(ContentFile<?> file) {
+    private boolean eval(ContentFile<?> file, boolean signedUuidMode) {
```
makes sense. this is a private method that we can always update later. We definitely want to include this for 1.11.0 release

stevenzwu commented on api/src/test/java/org/apache/iceberg/expressions/TestExpressionUtil.java:
```diff
@@ -1350,4 +1352,122 @@ private void assertEquals(UnboundTransform<?, ?> expected, UnboundTransform<?, ?
+    assertThat(result).isNotNull().isInstanceOf(UnboundPredicate.class);
+    UnboundPredicate<?> predicate = (UnboundPredicate<?>) result;
+    assertThat(predicate.literal().value()).isEqualTo(testUuid);
+    // The literal should now use the signed comparator
```
do we need to add an assertion after line 1379 that the literal's `useSignedComparator` var is `false`?

stevenzwu commented on api/src/test/java/org/apache/iceberg/expressions/TestStrictMetricsEvaluator.java:
```diff
@@ -684,4 +733,111 @@ SCHEMA, lessThanOrEqual("struct.nested_col_with_stats", INT_MAX_VALUE))
+  }
+
+  @Test
+  public void testStrictUuidLt() {
```
similar comment for testing other scenarios as above

stevenzwu commented on api/src/test/java/org/apache/iceberg/expressions/TestStrictMetricsEvaluator.java:
```diff
@@ -684,4 +733,111 @@ SCHEMA, lessThanOrEqual("struct.nested_col_with_stats", INT_MAX_VALUE))
+    assertThat(allMatch).as("Strict NOT IN should not match when value in bounds").isFalse();
+  }
+
+  // Tests for file with inverted UUID bounds (as would be written by legacy signed comparator)
```
nit: inverted -> signed

stevenzwu commented on api/src/test/java/org/apache/iceberg/types/TestComparators.java:
```diff
@@ -100,6 +100,35 @@ public void testUuid() {
         Comparators.forType(Types.UUIDType.get()),
         UUID.fromString("81873e7d-1374-4493-8e1d-9095eff7046c"),
         UUID.fromString("fd02441d-1423-4a3f-8785-c7dd5647e26b"));
+    assertComparesCorrectly(
```
do we also need to add coverage for `useSignedUuid` variation?

stevenzwu commented on data/src/test/java/org/apache/iceberg/data/TestMetricsRowGroupFilter.java:
```diff
@@ -1138,6 +1149,139 @@ public void testUUID() {
 
+  /**
+   * Tests UUID filtering with values that span the signed/unsigned comparison boundary. In RFC 9562
+   * unsigned comparison: UUID_LOW < UUID_MID < UUID_HIGH < UUID_HIGHER In legacy signed comparison:
```
nit: missing a `.` after `UUID_HIGHER`

stevenzwu commented on data/src/test/java/org/apache/iceberg/data/TestMetricsRowGroupFilter.java:
```diff
@@ -1138,6 +1149,139 @@ public void testUUID() {
+      records.add(record);
+    }
+
+    File parquetFile = writeParquetFile("uuid-high-only-test", UUID_SCHEMA, records);
```
this would still generate parquet column chunk stats using unsigned comparator?

Can we still generate parquet using the signed comparator to test Parquet files written in the legacy way?

stevenzwu reviewed (commented):
went through the unit test code. really appreciate the comprehensive coverage. awesome work!

bodduv commented on parquet/src/main/java/org/apache/iceberg/parquet/ParquetMetricsRowGroupFilter.java:
```diff
@@ -80,15 +103,20 @@ private class MetricsEvalVisitor extends BoundExpressionVisitor<Boolean> {
+    private boolean useSignedUuidComparator = false;
 
-    private boolean eval(MessageType fileSchema, BlockMetaData rowGroup) {
+    private boolean eval(MessageType fileSchema, BlockMetaData rowGroup, boolean signedUuidMode) {
```
I was impartial to this, but after your comment, I decided to have consistency among the three:
- `ParquetMetricsRowGroupFilter`
- `InclusiveMetricsEvaluator`
- `ManifestEvalutor`
Making the enclosed class static.

bodduv committed:
Address comments; Bring all three metrics evaluators in sync; improve tests

bodduv committed:
Resolving conflicts before merge

bodduv committed:
Updated branch from upstream main

bodduv committed:
Block UUID predicates in strict metrics evaluation

bodduv commented:
Prior to PR getting closed, the open issue was around how to handle `StrictMetricsEvaluator` for UUID comparisons. The uses of `StrictMetricsEvaluator` can be seen as an optimization, where data file metadata is used to make decisions where possible to avoid checking each row of the file. `StrictMetricsEvaluator` returns `ROWS_MIGHT_NOT_MATCH`, i.e., the current intended optimization could not be applied safely to the entire file without reading each of its rows.

In the recent commits post PR re-open, `StrictMetricsEvaluator` return `ROWS_MIGHT_NOT_MATCH` for UUID comparison (apart from `isNull` and `notNull` expressions) as its callers treat true as proof that every row in a file matches the predicate.

bodduv committed:
Tests: Add UUID strict metrics coverage

bodduv committed:
Merge remote-tracking branch 'origin/main' into uuid_comparison_bug

bodduv committed:
Updated branch from upstream master

# Conflicts:
#	parquet/src/main/java/org/apache/iceberg/parquet/ParquetMetricsRowGroupFilter.java

bodduv committed:
Tests: Update UUID delete coverage

bodduv commented:
Could I ask for another look @nastra @RussellSpitzer @stevenzwu @amogh-jahagirdar @pvary?
